### DISCUSSION TITLE
why pitch should be negated when transfer from euler to quaternion?

### DISCUSSION THREAD

**Discussion by Maxinjun** [novatel/novatel_oem7_driver#54] [Q&A]
Hi,
when i review the code ,I found the below code in ins_handler.cpp 

void publishImuMsg_OEM7(boost::shared_ptr<sensor_msgs::Imu>& imu)
    {
      if(inspva_)
      {
        tf2::Quaternion tf_orientation;
        tf_orientation.setRPY(
                           degreesToRadians(inspva_->roll),
                          -degreesToRadians(inspva_->pitch),
                          -degreesToRadians(inspva_->));
        imu->orientation = tf2::toMsg(tf_orientation);
      }
      else
      {
        ROS_WARN_THROTTLE(10, "INSPVA not available; 'Imu' message not generated.");
        return;
      }

according to the docment "OEM7_Commands_Logs_Manual_v14" page1002 ,roll、pitch angle are both right-handed, azimuth angle is left-handed, So Why there is  a '-' before pitch?  By the way,   is “oem7_imu_reference_frame_” in the code right(x), front(y), up(z) if I use default setting

**Comment by novatel-applications-engineering** [comment]
@Maxinjun Thank you for reaching out to us. 

The void publishImuMsg_OEM7() function is to provide backwards compatibility with the original implementation of the driver, which did not follow REP145. You should be using 'default' driver behavior instead, which is REP145 compliant. (just don't set the "oem7_imu_reference_frame" parameter) The pitch is negative, because the driver is following ROS convention (X-forward), vs OEM7 SPAN convention (in INSPVA), which is Y-Forward.

Are you observing anomalies in operation? 

Please let us know if you have any questions.
